Fix #21024 - Optimize x^^c expressions

faresbakhit opened the pull request:
- Avoid calls to `std.math.pow` for trivial exponents: x^^-1, x^^0, x^^1, x^^2.
- Add appropriate tests to `runnable/powinline.d` and `fail_compilation/powinline.d`.
- Add a field, `originalExp`, to `CommaExp`, to avoid cryptic error message involving the x^^2 rewrite.

If DMD supports a flag like `-ffast-math` in the future this rewrite should be more aggressive.

faresbakhit committed:
Fix #21024 - Optimize x^^c expressions

Reason for the *magic* constraint c<8 on inlining x^^c:

https://sourceware.org/git/?p=glibc.git;a=blob;f=sysdeps/x86_64/fpu/e_powl.S;h=47f129f34d368d7c67b8e5f2462b36b0bebb7621;hb=HEAD#l136

faresbakhit committed:
Fix poor assumption about expression state

faresbakhit commented:
This patch doesn't consider operator overloads for `^^`. Some questions:

- Are `x ^^ n` overloads required to match `x * ... * x (n times)`?
- The rewrites (`x ^^ 1.0` to `x`), (`x ^^ 0.0` to `1.0`), and (`x ^^ -1.0` to `1 / x`) also break because of operator overloading, how should those be handled?

The common-sense fix is to restrict the domain of `e` in `e^^c` to expressions of floating point type when applying this optimization.

dkorpel commented:
Yes I would only do this for built-in types.

faresbakhit committed:
Restrict optimization to floating point expressions

faresbakhit committed:
Generalize optimization to any scalar data type

faresbakhit committed:
Fix segfault on x^^c where x is a single member anonymous enum

DMD segfaulted on compiling the unittests in std/algorithm/sorting.o, the
unittest that caused the segfault can be reduced to:

    enum real Two = 2.0;
    auto _ = Two^^3;

I'm not sure why copying the anonymous enum into a `const` variable causes
the compiler to segfault.

faresbakhit committed:
Add tests to x^^c inlining optimization

faresbakhit commented:
Hi @dkorpel, can you take a look at this commit: https://github.com/faresbakhit/dmd/commit/d0b58c07523f017ebb71a7378760778ab9b2d146

Is this a bug in DMD?

faresbakhit committed:
Fix missing type for e1 ^^ -1 to 1 / e1 rewrite

dkorpel commented:
Possibly, have you checked with a debugger where the segfault happens? 

I do notice that you don't run `expressionSemantic` on the rewritten expression, so there is a good chance it breaks internal assumptions when it reaches code generation.

faresbakhit commented:
> I do notice that you don't run `expressionSemantic` on the rewritten expression, so there is a good chance it breaks internal assumptions when it reaches CTFE / code generation.

Yes, I relied on [`visit`](https://github.com/dlang/dmd/blob/02a64d2e1359119b91d2b932e61ed712f272507a/compiler/src/dmd/expressionsem.d#L12779) of [`ExpressionSemanticVisitor`](https://github.com/dlang/dmd/blob/02a64d2e1359119b91d2b932e61ed712f272507a/compiler/src/dmd/expressionsem.d#L3812) to add semantic information after the optimization, but after reading [`expressionSemantic`](https://github.com/dlang/dmd/blob/stable/compiler/src/dmd/expressionsem.d#L13882) now, I realize that it doesn't work like that.

https://github.com/dlang/dmd/blob/02a64d2e1359119b91d2b932e61ed712f272507a/compiler/src/dmd/expressionsem.d#L12790-L12797

faresbakhit commented:
The tests [`fail_compilation/fail19319a.d`](https://github.com/dlang/dmd/blob/02a64d2e1359119b91d2b932e61ed712f272507a/compiler/test/fail_compilation/fail19319a.d) and [`fail_compilation/fail19319b.d`](https://github.com/dlang/dmd/blob/stable/compiler/test/fail_compilation/fail19319b.d) do not fail now because `x^^7` is now inlined and don't need `std.math`. Should `7` be changed to another number that does actually invoke the (requires `std.math` for `^^` operators) error?

dkorpel commented:
Either that, or the optimization shouldn't be done for larger exponents.

I honestly don't think we should be going up to 8 here. The original issue suggested short-circuiting common obvious cases, so those don't depend on `std.math`. Generating faster code for integer exponents other than the trivial ones seems more like a job for the optimizer to me. Looking at GCC and clang, they can always optimize `powf(x, -1)`, `powf(x, 0)`, `powf(x, 1)` and `powf(x, 2)`, but require `-ffast-math` to go below -1 and above 2: https://godbolt.org/z/EsdsWvz8s

That suggests that rewriting `powf(x, 3)` to `x*x*x` is not IEEE compliant. While we don't have to define `^^` to be equivalent to the standard C function, I don't like the added complexity of the frontend trying to pre-optimize `x ^^ 7` either way.

@ibuclaw since you opened the issue, what do you think?

faresbakhit committed:
Move rewrites from constant folding to expression semantic and restrict them to [-1, 2]

faresbakhit commented:
I think it's correct that the rewrite is not IEEE complaint for c not in [-1, 2]; one of the unittests[^1] in [libmir/mir-random](https://github.com/libmir/mir-random) appears to be failing because of it. The rewrite shouldn't be a problem for non-floating point scalar expressions, though.

On another note, the rewrite causes _bad_ errors, they are not much of a problem for x^^{-1,0,1} as they are for 2 (or greater if we inline non-floating point scalars exponentiation), for example:

```
ex.d(4): Error: cannot implicitly convert expression `(const const(double) __powtmp2 = x + 5.0;) , __powtmp2 * ...` of type `double` to `int`
    int y = (x + 5)^^2;
            ^
```

[^1]: https://github.com/libmir/mir-random/blob/master/source/mir/random/flex/package.d#L875

faresbakhit committed:
Improve error message for the x^^2 rewrite.

Before:
    ex.d(4): Error: can implicitly convert expression `(const const(double) __powtmp2 = x + 5.0;) , __powtmp2 * ...` of type `double` to `int`
        int y = ( x + 5 ) ^^ 2;
                ^
and after:
    ex.d(4): Error: cannot implicitly convert expression `(x + 5.0) ^^ 2L` of type `double` to `int`
        int y = ( x + 5 ) ^^ 2;
                ^

faresbakhit commented:
I think this is ready for review.

CC @dkorpel

faresbakhit committed:
Update C++ frontend header to match change in `CommaExp`

dkorpel commented on compiler/src/dmd/hdrgen.d:
```diff
@@ -2724,6 +2724,12 @@ private void expressionPrettyPrint(Expression e, ref OutBuffer buf, ref HdrGenSt
     {
+        if (e.originalExp !is null)
+        {
+            e.originalExp.expressionPrettyPrint(buf, hgs);
```
Can you add a test case for this, by adding an error related to `x ^^ 2` to a file in fail_compilation that prints an error message with it? You can use an existing test file as a template for getting a TEST_OUTPUT block, and run `rdmd run.d your_test_name.d AUTO_UPDATE=1` to automatically update the TEST_OUTPUT block.

faresbakhit replied:
Addressed in the last commit, https://github.com/dlang/dmd/pull/21082/commits/72754cdebc284e7d87d8dd693d5b2d41a676fdf7.

dkorpel commented on compiler/test/runnable/powinline.d:
```diff
@@ -0,0 +1,44 @@
+/*
+REQUIRED_ARGS: -O
```
This switch is not necessary, since this optimization is done in the frontend, while that switch only affects the backend. Instead, maybe you can put the `-betterC` switch in there, to ensure the rewritten ^^ expressions don't rely on Phobos / druntime  anymore.

faresbakhit replied:
Addressed in the last commit, https://github.com/dlang/dmd/pull/21082/commits/72754cdebc284e7d87d8dd693d5b2d41a676fdf7.

dkorpel commented on compiler/test/runnable/powinline.d:
```diff
@@ -0,0 +1,44 @@
+    static assert(4.0 ^^ -1 == 0.25);
+}
+
+void main()
```
```suggestion
extern(C) void main()
```

In case my `-betterC` suggestion works.

faresbakhit replied:
Addressed in the last commit, https://github.com/dlang/dmd/pull/21082/commits/72754cdebc284e7d87d8dd693d5b2d41a676fdf7.

dkorpel commented on compiler/src/dmd/expressionsem.d:
```diff
@@ -12796,6 +12796,59 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
+            if (expo == -1)
+            {
+                Expression ex = new DivExp(exp.loc, one, be.e1);
+                ex = ex.optimize(WANTvalue);
```
Is this constant folding here necessary? In theory that should be handled after expression semantic, but maybe you encountered a test failure?

faresbakhit replied:
I think in an before moving the code to semantic analysis const expressions weren't folded when I inspected `-vcg-ast` but testing now, it const folds just fine.

This line is unnecessary.

faresbakhit replied:
Addressed in the last commit, https://github.com/dlang/dmd/pull/21082/commits/72754cdebc284e7d87d8dd693d5b2d41a676fdf7.

dkorpel commented on compiler/src/dmd/expressionsem.d:
```diff
@@ -12796,6 +12796,59 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         }
 
+        // Inline the expression, if possible.
+        BinExp be = cast(BinExp)e;
```
Since `PowExp` is a subclass of `BinExp`, is there a reason to explicitly cast it to `be` instead of using `exp.e1` and `exp.e2` directly?

faresbakhit replied:
I didn't use `exp.e1` and `exp.e2` because `optimize` converts `e2` to an `IntegerExp` (if possible) that we check for in the second condition in `be.e1.type.isScalar() && be.e2.isIntegerExp()`.

I didn't like the cast so I thought that instead of casting to `PowExp`, I cast to the lowest common denominator that has `e1` and `e2` (`BinExp`) but thinking about it now this is just bad analysis on my part.

faresbakhit replied:
Addressed in the last commit, https://github.com/dlang/dmd/pull/21082/commits/72754cdebc284e7d87d8dd693d5b2d41a676fdf7.

Is it correct, that way?

dkorpel replied:
Oh, I just noticed that `e` is different from `exp` at that point because it's the return value of `exp.optimize`. In that case either a cast to BinExp or PowExp is fine.

dkorpel commented on compiler/src/dmd/expressionsem.d:
```diff
@@ -12796,6 +12796,59 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
+            // Replace e1 ^^ 2 with e1 * e1
+            else if (expo == 2)
+            {
+                auto v = copyToTemp(STC.const_, "__powtmp", be.e1);
```
Good that you thought about side-effects here :+1:

faresbakhit replied:
Thanks!

dkorpel reviewed (commented):
Pretty good overall, just some minor things. Only the .h file update is critical.

faresbakhit committed:
Address code review feedback

Co-authored-by: Dennis Korpel <dkorpel@gmail.com>

faresbakhit commented on compiler/src/dmd/frontend.h:
```diff
@@ -2766,6 +2766,7 @@ class CommaExp final : public BinExp
 public:
     const bool isGenerated;
     bool allowCommaExp;
+    Expression* originalExp;
```
Addressed in the last commit, https://github.com/dlang/dmd/pull/21082/commits/72754cdebc284e7d87d8dd693d5b2d41a676fdf7.
